### PR TITLE
Add script to update & verify repo-infra when subtree'ed in

### PR DESCRIPTION
**Pull request by hoegaarden**
This could be a way to keep `repo-infra` up to date in other repos which have `subtree add`ed `repo-infra`, see also https://github.com/kubernetes/repo-infra/pull/82#issuecomment-452683686.

/cc @nikhita 
/cc @tao12345666333 
/cc @ixdy 
/cc @mariantalla

### CODE REVIEW

**Comment by tao12345666333** [comment]
Thanks for notifying me. :heart:

File: verify/test-repo-infra-subtree.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eu
+set -o pipefail
+
+readonly WORKSPACE="$( mktemp -d )"
+trap 'rm -rf -- "$WORKSPACE"' EXIT
+
+readonly REPO_INFRA_URL="$( git rev-parse --show-toplevel )"
```
Comment (tao12345666333) [inline verify/test-repo-infra-subtree.sh]: Maybe `REPO_INFRA_URL`  named like `REPO_INFRA_ROOT` or `REPO_INFRA_PATH` will more make sense?

**Comment by hoegaarden** [comment]
PTAL @tao12345666333

**Comment by tao12345666333** [comment]
@hoegaarden :+1:  Thanks.

**Comment by hoegaarden** [comment]
@ixdy @mikedanese @nikhita
Any input on that approach? Do you think that might be an appropriate mechanism to ensure and verify that the repo-infra is up2date in other repos?

File: verify/test-repo-infra-subtree.sh [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+set -o pipefail
+
+readonly WORKSPACE="$( mktemp -d )"
+trap 'rm -rf -- "$WORKSPACE"' EXIT
+
+[ "${TRAVIS:-}" = 'true' ] && {
+  git checkout "$BRANCH_TO_MERGE"
+}
+
+readonly REPO_INFRA_ROOT="$( git rev-parse --show-toplevel )"
```
Comment (ixdy) [inline verify/test-repo-infra-subtree.sh]: if this is running in another repo where it's been subtree'd, is this the root of that repo, or the root of that subtree?

**Comment by ixdy** [comment]
I guess one worry I have is that we already have at least 2 ways of handling third-party dependencies in several of our repos:
- go vendoring / go modules
- bazel workspace dependencies
- occasionally other language-specific methods (npm, requirements.txt, etc)

git subtrees might a good complimentary option, but I'd like other folks to chime in as well.

File: verify/test-repo-infra-subtree.sh (hunk 2 above)
Comment (hoegaarden) [reply verify/test-repo-infra-subtree.sh]: Note that this is "just" the test for the `{update,verify}-repo-infra-subtree.sh` scripts, it is supposed to run locally (while maybe hacking on `{update,verify}-repo-infra-subtree.sh` **and** in CI

The flow of the test is:
  1. create a temporary repo
  2. `subtree add` the current version (the PR branch) into the temporary repo
  3. run `{update,verify}-repo-infra-subtree.sh` in the temporary repo and check if they do what they promised

`$REPO_INFRA_ROOT` is supposed to be the local path to the git repo of the version of `k/repo-infra` that is under test / to be merged. `$REPO_INFRA_ROOT` is the path that will be used when the test [`subtree add`s](https://github.com/kubernetes/repo-infra/pull/94/files/b7b803d1efca6335e42311935df0ed2fe7cb9853#diff-06847f1dec8f17f07a568abdbcef5e06R71) `repo-infra` into the temporary repo.

I notice now that this test assumes to be run within the `repo-infra` directory. This just works fin in CI, but might break locally. I can see if I can guard against that for local invocations.

Note:
There was the request in [another review](https://github.com/kubernetes/repo-infra/pull/94#pullrequestreview-191674251) to rename that variable from `REPO_INFRA_URL` to `REPO_INFRA_ROOT` which might make more sense in the context of the test. Or maybe it doesn't. I am bad at naming.

**Comment by hoegaarden** [comment]
> I guess one worry I have is that we already have at least 2 ways of handling third-party dependencies in several of our repos:

I just learned about `git subtree` from this repo, it is specifically stated in [this repo's README][rm] which led me to believe that this method is already a thing.

> [...] but I'd like other folks to chime in as well.

Other folks don't care ... ;)

[rm]: https://github.com/kubernetes/repo-infra/blob/a3f1f8d5b18ad754f553bb59a80334f4f70405a6/README.md#using-git-subtree

**Comment by fejta** [comment]
Travis is going away, and it is far from clear that subtrees are something we want to use widely. I would recommend closing this PR until there is more evidence that repos are finding subtrees a great and amazing pattern others should use.

**Pull request closed without merging**